grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java

```diff
@@ -0,0 +1,128 @@
+  interface RequestInfo {
+
+    String token();
+
+    Map<String, String> headers();
+
+    @Nullable
+    SocketAddress remoteAddress();
+  }
+
+  public NewConnectionInterceptor(
+      Persistence persistence, AuthenticationService authenticationService) {
+    this.persistence = persistence;
+    this.authenticationService = authenticationService;
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    try {
+      String token = headers.get(TOKEN_KEY);
+      if (token == null) {
+        call.close(Status.UNAUTHENTICATED.withDescription("No token provided"), new Metadata());
+        return new NopListener<>();
+      }
+
+      RequestInfo info =
+          ImmutableRequestInfo.builder()
+              .token(token)
+              .headers(convertAndFilterHeaders(headers))
+              .remoteAddress(call.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR))
+              .build();
+
+      Connection connection = connectionCache.get(info);
+
+      Context context = Context.current();
+      context = context.withValue(GrpcService.CONNECTION_KEY, connection);
+      return Contexts.interceptCall(context, call, headers, next);
+    } catch (Exception e) {
+      call.close(
+          Status.UNAUTHENTICATED.withDescription("Invalid token").withCause(e), new Metadata());
+    }
+    return new NopListener<>();
+  }
+
+  private Connection newConnection(RequestInfo info) throws UnauthorizedException {
+    AuthenticationSubject authenticationSubject = authenticationService.validateToken(info.token());
+
+    AuthenticatedUser user = authenticationSubject.asUser();
+    Connection connection;
+    if (!user.isFromExternalAuth()) {
+      SocketAddress remoteAddress = info.remoteAddress();
+      InetSocketAddress inetSocketAddress = DUMMY_ADDRESS;
+      if (remoteAddress instanceof InetSocketAddress) {
+        inetSocketAddress = (InetSocketAddress) remoteAddress;
+      }
+      connection = persistence.newConnection(new ClientInfo(inetSocketAddress, null));
+    } else {
+      connection = persistence.newConnection();
+    }
+    connection.login(user);
+    if (user.token() != null) {
+      connection.clientInfo().ifPresent(c -> c.setAuthenticatedUser(user));
+    }
+    connection.setCustomProperties(info.headers());
+    return connection;
+  }
+
+  private static Map<String, String> convertAndFilterHeaders(Metadata headers) {
+    Map<String, String> stringHeaders = new HashMap<>();
+    for (String key : headers.keys()) {
+      // Ignore gRPC specific (in addition to binary) headers because they're not used by Connection
+      // and some of the values contain unique values for each request which prevents effective
+      // caching.
+      if (key.endsWith("-bin") || key.startsWith("grpc-")) {
+        continue;
+      }
+      String value = headers.get(Key.of(key, Metadata.ASCII_STRING_MARSHALLER));
```

Comment:
I know this is from old code, but isn't there an option to iterate key/value pairs instead of pulling the value here?

Reply:
It looks like `Metadata` doesn't implement `Map<K,V>` and doesn't have that funcationality.